src/merkle_tree.rs

```diff
@@ -0,0 +1,470 @@
+/// data elements to the leaves level so the amount is a power of two
+pub fn create_merkle_tree_from_data(inputs: &Input) -> Result<MerkleTree, MerkleTreeError> {
+    if inputs.is_empty() {
+        return Err(MerkleTreeError::DataLengthIsZero);
+    }
+    let mut merkle_tree = MerkleTree::default();
+    let data_hashes_level = create_data_hashes(inputs);
+
+    let tree = create_hash_tree_from_leaves(data_hashes_level);
+
+    merkle_tree.merkle_tree = tree;
+    merkle_tree.set_last_element_index(inputs.len() - 1);
+    Ok(merkle_tree)
+}
+
+/// Allows the creation of a Merkle Tree from its base level (or leaves level) which are the hashes of the original
+/// data. This function is useful to extend the Merkle Tree when adding a new element.
+fn create_hash_tree_from_leaves(leaves: Vec<String>) -> HashTree {
+    let mut merkle_tree = Vec::new();
+    merkle_tree.push(leaves.clone());
+    let mut previous_level = leaves;
+    let mut level_len = previous_level.len();
+
+    while level_len != 1 {
+        let new_level = create_new_level(&previous_level);
+        merkle_tree.insert(0, new_level.clone());
+        previous_level = new_level;
+        level_len = previous_level.len();
+    }
+    merkle_tree
+}
+
+/// Creates the leaves level from a specific input. Meaning it hashes every data element from the input and returns
+/// a vector of those hashes that can later be used as the lower level of the tree.
+/// Notice that it fills with default elements if necessary to reach an amount that is a power of two.
+fn create_data_hashes(inputs: &Input) -> Vec<String> {
+    let mut data_hashes = Vec::new();
+
+    for i in inputs {
+        data_hashes.push(sha256::digest(i));
+    }
+    let data_len = inputs.len();
+    if !data_len.is_power_of_two() {
+        let fill_amount = data_len.next_power_of_two() - data_len;
+        data_hashes.append(&mut vec![
+            DEFAULT_ZERO_HASH.to_string().clone();
+            fill_amount
+        ]);
+    }
+
+    data_hashes
+}
+
+/// Given the previous constructed level of the Merkle Tree it allows to create a parent level for it by concatenating
+/// elements in pairs and hashing the results. It is helpful to construct a Merkle Tree from bottom to top.
+fn create_new_level(previous_level: &[String]) -> Vec<String> {
+    let mut index = 0;
+    let mut new_level = Vec::new();
+
+    while index < previous_level.len() {
+        let element_concat = previous_level[index].clone() + &previous_level[index + 1];
+        let element_hash = sha256::digest(element_concat);
+        new_level.push(element_hash);
+        index += 2;
+    }
+
+    new_level
+}
+
+/// Given a proof provided by the Merkle Tree it allows to verify that a certain element has been stored at a certain index
+/// in the Merkle Tree.
+pub fn verify_element(
+    element_data: Vec<u8>,
+    proof: &Vec<String>,
+    root: String,
+    element_index: usize,
+) -> bool {
```

Comment:
Here using a slice allows users to pass both `Vec<String>` or `[String; N]` to the function.

```suggestion
pub fn verify_element(
    element_data: Vec<u8>,
    proof: &[String],
    root: String,
    element_index: usize,
) -> bool {
```

Reply:
Fixed! Thank you :)